Index: server/physics/Body.hh
===================================================================
--- server/physics/Body.hh	(revision 7891)
+++ server/physics/Body.hh	(working copy)
@@ -259,6 +259,8 @@
     protected: ParamT<double> *iyzP;
     protected: bool customMassMatrix;
     protected: double cx,cy,cz,bodyMass,ixx,iyy,izz,ixy,ixz,iyz;
+
+    private: std::string lightName;
   };
 
   /// \}
Index: server/physics/Geom.cc
===================================================================
--- server/physics/Geom.cc	(revision 7891)
+++ server/physics/Geom.cc	(working copy)
@@ -250,7 +250,7 @@
 
   if (this->placeable && !this->IsStatic())
   {
-    if (dGeomGetClass(geomId) != dTriMeshClass)
+    //if (dGeomGetClass(geomId) != dTriMeshClass)
     {
       this->transId = dCreateGeomTransform( this->spaceId );
       dGeomTransformSetGeom( this->transId, this->geomId );
Index: server/physics/Body.cc
===================================================================
--- server/physics/Body.cc	(revision 7891)
+++ server/physics/Body.cc	(working copy)
@@ -242,6 +242,10 @@
     }
   }
 
+  if ((childNode = node->GetChild("light")))
+  {
+    this->lightName = OgreCreator::CreateLight(childNode, this->cgVisual);
+  }
 
   childNode = node->GetChildByNSPrefix("sensor");
 
Index: worlds/bandit.world
===================================================================
--- worlds/bandit.world	(revision 7891)
+++ worlds/bandit.world	(working copy)
@@ -16,7 +16,7 @@
   <verbosity>5</verbosity>
 
   <physics:ode>
-    <stepTime>0.03</stepTime>
+    <stepTime>0.01</stepTime>
     <gravity>0 0 -9.8</gravity>
     <cfm>10e-5</cfm>
     <erp>0.3</erp>
@@ -58,7 +58,7 @@
     </include>
 
     <model:physical name="box_model">
-      <xyz>0.15 0 0.125</xyz>
+      <xyz>0.05 0 0.125</xyz>
       <static>false</static>
 
       <attach>
Index: worlds/test_stacks_with_rays.world
===================================================================
--- worlds/test_stacks_with_rays.world	(revision 7891)
+++ worlds/test_stacks_with_rays.world	(working copy)
@@ -28,7 +28,7 @@
     <cfm>0.0000000001</cfm>
     <erp>0.2</erp>
     <quickStep>true</quickStep>
-    <quickStepIters>10</quickStepIters>
+    <quickStepIters>100</quickStepIters>
     <quickStepW>1.3</quickStepW>
   </physics:ode>
 
Index: worlds/test_stacks.world
===================================================================
--- worlds/test_stacks.world	(revision 7891)
+++ worlds/test_stacks.world	(working copy)
@@ -28,7 +28,7 @@
     <cfm>0.0000000001</cfm>
     <erp>0.1</erp>
     <quickStep>true</quickStep>
-    <quickStepIters>30</quickStepIters>
+    <quickStepIters>20</quickStepIters>
     <quickStepW>1.3</quickStepW>
   </physics:ode>
 
Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(revision 7891)
+++ CMakeLists.txt	(working copy)
@@ -30,7 +30,7 @@
 SET (gazebocontroller_sources_desc "List of controller sources" 
      CACHE INTERNAL "Gazebo controller sources list description" FORCE)
 
-SET (OGRE_VERSION 1.6.2 CACHE INTERNAL "Ogre version requirement" FORCE)
+SET (OGRE_VERSION 1.6.1 CACHE INTERNAL "Ogre version requirement" FORCE)
 
 SET (FREEIMAGE_MAJOR_VERSION 3 CACHE INTERNAL "FreeImage major version requirement" FORCE)
 SET (FREEIMAGE_MINOR_VERSION 10 CACHE INTERNAL "FreeImage minor version requirement" FORCE)
