Index: libgazebo/Client.cc
===================================================================
--- libgazebo/Client.cc	(revision 8187)
+++ libgazebo/Client.cc	(working copy)
@@ -172,7 +172,7 @@
     catch (std::string e)
     {
       std::cerr << "Error Opening SimulationIface [" << e << "]\n";
-      //exit(0); // don't exit, retry
+      exit(0);
     }
   }
 
Index: server/rendering/Image.cc
===================================================================
--- server/rendering/Image.cc	(revision 8187)
+++ server/rendering/Image.cc	(working copy)
@@ -174,7 +174,8 @@
       return clr;
     }
 
-    if (FREEIMAGE_COLORORDER == FREEIMAGE_COLORORDER_RGB)
+    //if (FREEIMAGE_COLORORDER == FREEIMAGE_COLORORDER_RGB) /// @todo: freeimage 3.9.3 does not have these defined
+    if (1)
     {
       clr.Set(  firgb.rgbRed, firgb.rgbGreen, firgb.rgbBlue);
     }
Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(revision 8187)
+++ CMakeLists.txt	(working copy)
@@ -34,7 +34,7 @@
 SET (OGRE_VERSION 1.6.1 CACHE INTERNAL "Ogre version requirement" FORCE)
 
 SET (FREEIMAGE_MAJOR_VERSION 3 CACHE INTERNAL "FreeImage major version requirement" FORCE)
-SET (FREEIMAGE_MINOR_VERSION 10 CACHE INTERNAL "FreeImage minor version requirement" FORCE)
+SET (FREEIMAGE_MINOR_VERSION 9 CACHE INTERNAL "FreeImage minor version requirement" FORCE)
 SET (FREEIMAGE_VERSION ${FREEIMAGE_MAJOR_VERSION}.${FREEIMAGE_MINOR_VERSION}.0 CACHE INTERNAL "FreeImage version requirement" FORCE)
 SET (ODE_VERSION 0.10.1 CACHE INTERNAL "ODE version requirement" FORCE)
 SET (MIN_BOOST_VERSION 1.35.0 CACHE INTERNAL "Boost min version requirement" FORCE)
