Index: server/physics/Body.hh
===================================================================
--- server/physics/Body.hh	(revision 7929)
+++ server/physics/Body.hh	(working copy)
@@ -268,6 +268,8 @@
     protected: ParamT<double> *iyzP;
     protected: bool customMassMatrix;
     protected: double cx,cy,cz,bodyMass,ixx,iyy,izz,ixy,ixz,iyz;
+
+    private: std::string lightName;
   };
 
   /// \}
Index: server/physics/Geom.cc
===================================================================
--- server/physics/Geom.cc	(revision 7929)
+++ server/physics/Geom.cc	(working copy)
@@ -249,7 +249,7 @@
 
   if (this->placeable && !this->IsStatic())
   {
-    if (dGeomGetClass(geomId) != dTriMeshClass)
+    //if (dGeomGetClass(geomId) != dTriMeshClass)
     {
       this->transId = dCreateGeomTransform( this->spaceId );
       dGeomTransformSetGeom( this->transId, this->geomId );
Index: server/physics/Body.cc
===================================================================
--- server/physics/Body.cc	(revision 7929)
+++ server/physics/Body.cc	(working copy)
@@ -242,6 +242,10 @@
     }
   }
 
+  if ((childNode = node->GetChild("light")))
+  {
+    this->lightName = OgreCreator::CreateLight(childNode, this->cgVisual);
+  }
 
   childNode = node->GetChildByNSPrefix("sensor");
 
Index: server/physics/ode/ODEPhysics.cc
===================================================================
--- server/physics/ode/ODEPhysics.cc	(revision 7929)
+++ server/physics/ode/ODEPhysics.cc	(working copy)
@@ -66,10 +66,10 @@
   this->contactGroup = dJointGroupCreate(0);
 
   // Help prevent "popping of deeply embedded object
-  dWorldSetContactMaxCorrectingVel(this->worldId, 10.0);
+  dWorldSetContactMaxCorrectingVel(this->worldId, 1.0);
 
   // This helps prevent jittering problems.
-  dWorldSetContactSurfaceLayer(this->worldId, 0.01);
+  dWorldSetContactSurfaceLayer(this->worldId, 0.001);
 
   Param::Begin(&this->parameters);
   this->globalCFMP = new ParamT<double>("cfm", 10e-5, 0);
Index: worlds/bandit.world
===================================================================
--- worlds/bandit.world	(revision 7929)
+++ worlds/bandit.world	(working copy)
@@ -16,7 +16,7 @@
   <verbosity>5</verbosity>
 
   <physics:ode>
-    <stepTime>0.03</stepTime>
+    <stepTime>0.01</stepTime>
     <gravity>0 0 -9.8</gravity>
     <cfm>10e-5</cfm>
     <erp>0.3</erp>
@@ -58,7 +58,7 @@
     </include>
 
     <model:physical name="box_model">
-      <xyz>0.15 0 0.125</xyz>
+      <xyz>0.05 0 0.125</xyz>
       <static>false</static>
 
       <attach>
Index: worlds/test_stacks.world
===================================================================
--- worlds/test_stacks.world	(revision 7929)
+++ worlds/test_stacks.world	(working copy)
@@ -25,10 +25,10 @@
   <physics:ode>
     <stepTime>0.001</stepTime>
     <gravity>0 0 -9.8</gravity>
-    <cfm>0.0000000001</cfm>
+    <cfm>0.001</cfm>
     <erp>0.1</erp>
     <quickStep>true</quickStep>
-    <quickStepIters>30</quickStepIters>
+    <quickStepIters>200</quickStepIters>
     <quickStepW>1.3</quickStepW>
   </physics:ode>
 
@@ -65,7 +65,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -85,7 +88,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -105,7 +111,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -125,7 +134,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -145,7 +157,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -165,7 +180,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -185,7 +203,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -205,7 +226,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -225,7 +249,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -245,7 +272,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -265,7 +295,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -285,7 +318,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -305,7 +341,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -325,7 +364,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -345,7 +387,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -365,7 +410,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -385,7 +433,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -405,7 +456,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -425,7 +479,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -445,7 +502,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -465,7 +525,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -485,7 +548,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -505,7 +571,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -525,7 +594,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -545,7 +617,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -565,7 +640,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -585,7 +663,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -605,7 +686,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -625,7 +709,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -645,7 +732,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -665,7 +755,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -685,7 +778,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -705,7 +801,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -725,7 +824,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -745,7 +847,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -765,7 +870,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -785,7 +893,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -805,7 +916,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -825,7 +939,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -845,7 +962,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -865,7 +985,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -885,7 +1008,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -905,7 +1031,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -925,7 +1054,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -945,7 +1077,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -965,7 +1100,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -985,7 +1123,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1005,7 +1146,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1025,7 +1169,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1045,7 +1192,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1065,7 +1215,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1085,7 +1238,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1105,7 +1261,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1125,7 +1284,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1145,7 +1307,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1165,7 +1330,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1185,7 +1353,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1205,7 +1376,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1225,7 +1399,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1245,7 +1422,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1265,7 +1445,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1285,7 +1468,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1305,7 +1491,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1325,7 +1514,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1345,7 +1537,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1365,7 +1560,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1385,7 +1583,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1405,7 +1606,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1425,7 +1629,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1445,7 +1652,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1465,7 +1675,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1485,7 +1698,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1505,7 +1721,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1525,7 +1744,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1545,7 +1767,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1565,7 +1790,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1585,7 +1813,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1605,7 +1836,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1625,7 +1859,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1645,7 +1882,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1665,7 +1905,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1685,7 +1928,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1705,7 +1951,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1725,7 +1974,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1745,7 +1997,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1765,7 +2020,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1785,7 +2043,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1805,7 +2066,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1825,7 +2089,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1845,7 +2112,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1865,7 +2135,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1885,7 +2158,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1905,7 +2181,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1925,7 +2204,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1945,7 +2227,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1965,7 +2250,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -1985,7 +2273,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -2005,7 +2296,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -2025,7 +2319,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -2045,7 +2342,10 @@
         <size>0.2</size>
         <mass>0.0</mass>
 
-        <mu1>109999.0</mu1>
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
 
         <visual>
           <size>0.4 0.4 0.4</size>
@@ -2066,6 +2366,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2084,6 +2389,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2102,6 +2412,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2120,6 +2435,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2138,6 +2458,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2156,6 +2481,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2174,6 +2504,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2192,6 +2527,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2210,6 +2550,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2228,6 +2573,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2246,6 +2596,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2264,6 +2619,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2282,6 +2642,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2300,6 +2665,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2318,6 +2688,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2336,6 +2711,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2354,6 +2734,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2372,6 +2757,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2390,6 +2780,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2408,6 +2803,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2426,6 +2826,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2444,6 +2849,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2462,6 +2872,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2480,6 +2895,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2498,6 +2918,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2516,6 +2941,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2534,6 +2964,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2552,6 +2987,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2570,6 +3010,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2588,6 +3033,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2606,6 +3056,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2624,6 +3079,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2642,6 +3102,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2660,6 +3125,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2678,6 +3148,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2696,6 +3171,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2714,6 +3194,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2732,6 +3217,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2750,6 +3240,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2768,6 +3263,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2786,6 +3286,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2804,6 +3309,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2822,6 +3332,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2840,6 +3355,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2858,6 +3378,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2876,6 +3401,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2894,6 +3424,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2912,6 +3447,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2930,6 +3470,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2948,6 +3493,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2966,6 +3516,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -2984,6 +3539,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3002,6 +3562,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3020,6 +3585,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3038,6 +3608,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3056,6 +3631,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3074,6 +3654,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3092,6 +3677,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3110,6 +3700,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3128,6 +3723,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3146,6 +3746,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3164,6 +3769,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3182,6 +3792,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3200,6 +3815,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3218,6 +3838,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3236,6 +3861,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3254,6 +3884,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3272,6 +3907,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3290,6 +3930,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3308,6 +3953,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3326,6 +3976,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3344,6 +3999,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3362,6 +4022,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3380,6 +4045,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3398,6 +4068,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3416,6 +4091,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3434,6 +4114,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3452,6 +4137,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3470,6 +4160,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3488,6 +4183,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3506,6 +4206,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3524,6 +4229,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3542,6 +4252,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3560,6 +4275,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3578,6 +4298,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3596,6 +4321,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3614,6 +4344,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3632,6 +4367,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3650,6 +4390,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3668,6 +4413,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3686,6 +4436,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3704,6 +4459,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3722,6 +4482,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3740,6 +4505,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3758,6 +4528,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3776,6 +4551,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3794,6 +4574,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3812,6 +4597,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3830,6 +4620,11 @@
         <size>0.5 0.5 1.0</size>
         <mass>0.1</mass>
 
+        <mu1>10.0</mu1>
+        <mu2>10.0</mu2>
+        <kp>100000000.0</kp>
+        <kd>1.0</kd>
+
         <visual>
           <size>0.5 0.5 1.0</size>
           <mesh>unit_box</mesh>
@@ -3870,7 +4665,7 @@
         <segments>10 10</segments>
         <uvTile>100 100</uvTile>
         <material>Gazebo/Rockwall</material>
-        <mu1>109999.0</mu1>
+        <mu1>1000.0</mu1>
         <mu2>1000.0</mu2>
       </geom:plane>
     </body:plane>
