Index: server/physics/Geom.cc
===================================================================
--- server/physics/Geom.cc	(revision 8080)
+++ server/physics/Geom.cc	(working copy)
@@ -249,12 +249,12 @@
 
   if (this->placeable && !this->IsStatic())
   {
-    if (dGeomGetClass(geomId) != dTriMeshClass)
+    //if (dGeomGetClass(geomId) != dTriMeshClass)
     {
       this->transId = dCreateGeomTransform( this->spaceId );
       dGeomTransformSetGeom( this->transId, this->geomId );
       dGeomTransformSetInfo( this->transId, 1 );
-      assert(dGeomGetSpace(this->geomId) == 0);
+      //assert(dGeomGetSpace(this->geomId) == 0);
     }
   }
   else if ( dGeomGetSpace(this->geomId) == 0 )
@@ -272,6 +272,7 @@
   }
   else
   {
+    // collide with all
     this->SetCategoryBits(GZ_ALL_COLLIDE);
     this->SetCollideBits(GZ_ALL_COLLIDE);
   }
