Index: server/rendering/UserCamera.cc
===================================================================
--- server/rendering/UserCamera.cc	(revision 7945)
+++ server/rendering/UserCamera.cc	(working copy)
@@ -41,7 +41,7 @@
 ////////////////////////////////////////////////////////////////////////////////
 /// Constructor
 UserCamera::UserCamera(GLWindow *parentWindow)
-  : OgreCamera("UserCamera")
+  : OgreCamera("UserCamera"),viewport( 0 )
 {
   std::stringstream stream;
 
@@ -126,7 +126,8 @@
 {
   this->window->resize(w, h);
   this->window->windowMovedOrResized();
-  this->viewport->setDimensions(0,0,1,1);
+  if (this->viewport)
+    this->viewport->setDimensions(0,0,1,1);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
