Index: libgazebo/Server.cc
===================================================================
--- libgazebo/Server.cc	(revision 8164)
+++ libgazebo/Server.cc	(working copy)
@@ -91,8 +91,6 @@
   stream << tmpdir << "/gazebo-" << user << "-" << this->serverId;
   this->filename = stream.str();
 
-  std::cout << "creating " << this->filename << "\n";
-
   // check to see if there is already a directory created.
   struct stat astat;
   if (stat(this->filename.c_str(), &astat) == 0) 
Index: libgazebo/Iface.cc
===================================================================
--- libgazebo/Iface.cc	(revision 8164)
+++ libgazebo/Iface.cc	(working copy)
@@ -190,14 +190,12 @@
   std::ios_base::fmtflags origFlags = std::cout.flags();
 
   // Print the name, version info
-  std::cout << "creating " << this->filename.c_str() << " "
+  // std::cout << "creating " << this->filename.c_str() << " "
+  // << setiosflags(std::ios::hex | std::ios::showbase)
+  // << std::setw(3) << ((GazeboData*) this->mMap)->version << " "
+  // << std::setiosflags(std::ios::dec | ~std::ios::showbase)
+  // << ((GazeboData*) this->mMap)->size << "\n";
 
-  << setiosflags(std::ios::hex | std::ios::showbase)
-  << std::setw(3) << ((GazeboData*) this->mMap)->version << " "
-
-  << std::setiosflags(std::ios::dec | ~std::ios::showbase)
-  << ((GazeboData*) this->mMap)->size << "\n";
-
   std::cout.flags(origFlags);
 
   this->Unlock();
