Index: server/physics/Geom.cc
===================================================================
--- server/physics/Geom.cc	(revision 8166)
+++ server/physics/Geom.cc	(working copy)
@@ -257,7 +257,7 @@
       this->transId = dCreateGeomTransform( this->spaceId );
       dGeomTransformSetGeom( this->transId, this->geomId );
       dGeomTransformSetInfo( this->transId, 1 );
-      assert(dGeomGetSpace(this->geomId) == 0);
+      //assert(dGeomGetSpace(this->geomId) == 0);
     }
   }
   else if ( dGeomGetSpace(this->geomId) == 0 )
