Index: server/physics/Body.hh
===================================================================
--- server/physics/Body.hh	(revision 7929)
+++ server/physics/Body.hh	(working copy)
@@ -268,6 +268,8 @@
     protected: ParamT<double> *iyzP;
     protected: bool customMassMatrix;
     protected: double cx,cy,cz,bodyMass,ixx,iyy,izz,ixy,ixz,iyz;
+
+    private: std::string lightName;
   };
 
   /// \}
Index: server/physics/Geom.cc
===================================================================
--- server/physics/Geom.cc	(revision 7929)
+++ server/physics/Geom.cc	(working copy)
@@ -249,7 +249,7 @@
 
   if (this->placeable && !this->IsStatic())
   {
-    if (dGeomGetClass(geomId) != dTriMeshClass)
+    //if (dGeomGetClass(geomId) != dTriMeshClass)
     {
       this->transId = dCreateGeomTransform( this->spaceId );
       dGeomTransformSetGeom( this->transId, this->geomId );
Index: server/physics/Body.cc
===================================================================
--- server/physics/Body.cc	(revision 7929)
+++ server/physics/Body.cc	(working copy)
@@ -242,6 +242,10 @@
     }
   }
 
+  if ((childNode = node->GetChild("light")))
+  {
+    this->lightName = OgreCreator::CreateLight(childNode, this->cgVisual);
+  }
 
   childNode = node->GetChildByNSPrefix("sensor");
 
Index: worlds/bandit.world
===================================================================
--- worlds/bandit.world	(revision 7929)
+++ worlds/bandit.world	(working copy)
@@ -16,7 +16,7 @@
   <verbosity>5</verbosity>
 
   <physics:ode>
-    <stepTime>0.03</stepTime>
+    <stepTime>0.01</stepTime>
     <gravity>0 0 -9.8</gravity>
     <cfm>10e-5</cfm>
     <erp>0.3</erp>
@@ -58,7 +58,7 @@
     </include>
 
     <model:physical name="box_model">
-      <xyz>0.15 0 0.125</xyz>
+      <xyz>0.05 0 0.125</xyz>
       <static>false</static>
 
       <attach>
