Index: cmake/SearchForStuff.cmake
===================================================================
--- cmake/SearchForStuff.cmake	(revision 8208)
+++ cmake/SearchForStuff.cmake	(working copy)
@@ -136,7 +136,7 @@
     SET (HAVE_FFMPEG TRUE)
   ENDIF (AVF_FOUND AND AVC_FOUND)
 
-  pkg_check_modules(PLAYER playercore>=3.0)
+  pkg_check_modules(PLAYER playercore>=2.2)
   IF (NOT PLAYER_FOUND)
     SET (INCLUDE_PLAYER OFF CACHE BOOL "Build gazebo plugin for player" FORCE)
     MESSAGE (STATUS "Warning: Player not found. The gazebo plugin for player will not be built. See the following website: http://playerstage.sourceforge.net")
Index: libgazebo/Iface.cc
===================================================================
--- libgazebo/Iface.cc	(revision 8208)
+++ libgazebo/Iface.cc	(working copy)
@@ -288,6 +288,12 @@
   // Map the mmap file
   this->mMap = mmap(0, this->size, PROT_READ | PROT_WRITE, MAP_SHARED, this->mmapFd, 0);
 
+  // flushes changes made to the in-core copy of a mmap file
+  if (msync( this->mMap, this->size, MS_SYNC))
+  {
+    stream << "error msync mmap: " << strerror(errno);
+    throw(stream.str());
+  }
 
   if (this->mMap == MAP_FAILED)
   {
Index: server/Simulator.cc
===================================================================
--- server/Simulator.cc	(revision 8208)
+++ server/Simulator.cc	(working copy)
@@ -358,7 +358,7 @@
       }
 
       World::Instance()->ProcessEntitiesToLoad();
-      World::Instance()->GraphicsUpdate();
+      //World::Instance()->GraphicsUpdate();
 
       currTime = this->GetWallTime();
 
Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(revision 8208)
+++ CMakeLists.txt	(working copy)
@@ -34,7 +34,7 @@
 SET (OGRE_VERSION 1.6.1 CACHE INTERNAL "Ogre version requirement" FORCE)
 
 SET (FREEIMAGE_MAJOR_VERSION 3 CACHE INTERNAL "FreeImage major version requirement" FORCE)
-SET (FREEIMAGE_MINOR_VERSION 10 CACHE INTERNAL "FreeImage minor version requirement" FORCE)
+SET (FREEIMAGE_MINOR_VERSION 9 CACHE INTERNAL "FreeImage minor version requirement" FORCE)
 SET (FREEIMAGE_VERSION ${FREEIMAGE_MAJOR_VERSION}.${FREEIMAGE_MINOR_VERSION}.0 CACHE INTERNAL "FreeImage version requirement" FORCE)
 SET (ODE_VERSION 0.10.1 CACHE INTERNAL "ODE version requirement" FORCE)
 SET (MIN_BOOST_VERSION 1.35.0 CACHE INTERNAL "Boost min version requirement" FORCE)
