--- opencv-1.0.0/otherlibs/highgui/cvcap_ffmpeg.cpp.orig	2008-07-30 18:49:18.702001218 -0700
+++ opencv-1.0.0/otherlibs/highgui/cvcap_ffmpeg.cpp	2008-07-30 18:50:24.254821875 -0700
@@ -38,10 +38,12 @@
 // the use of this software, even if advised of the possibility of such damage.
 //
 //M*/
 
 #include "_highgui.h"
+#define INT64_C
+#define __STDC_CONSTANT_MACROS
 
 
 extern "C" {
 #include <ffmpeg/avformat.h>
 }
@@ -779,11 +781,11 @@
 		av_freep(&mywriter->oc->streams[i]);
 	}
 
 	if (!(mywriter->fmt->flags & AVFMT_NOFILE)) {
 		/* close the output file */
-		url_fclose(&mywriter->oc->pb);
+		url_fclose(mywriter->oc->pb);
 	}
 
 	/* free the stream */
 	av_free(mywriter->oc);
 
--- opencv-1.0.0/interfaces/swig/python/Makefile.in.orig	2008-07-30 18:49:05.175160982 -0700
+++ opencv-1.0.0/interfaces/swig/python/Makefile.in	2008-07-30 18:51:23.188863477 -0700
@@ -345,11 +345,11 @@
 @BUILD_PYTHON_WRAPPERS_TRUE@                         highgui.i
 
 @BUILD_PYTHON_WRAPPERS_TRUE@_highgui_la_CXXFLAGS = $(PYTHON_CSPEC)
 @BUILD_PYTHON_WRAPPERS_TRUE@_highgui_la_LIBADD = $(top_builddir)/otherlibs/highgui/libhighgui.la
 @BUILD_PYTHON_WRAPPERS_TRUE@_highgui_la_LDFLAGS = -module -avoid-version -no-undefined $(PYTHON_LSPEC) \
-@BUILD_PYTHON_WRAPPERS_TRUE@                           -L$(SWIG_PYTHON_LIBS)
+@BUILD_PYTHON_WRAPPERS_TRUE@                           -L$(SWIG_PYTHON_LIBS) -Wl,-rpath,`rospack find ffmpeg`/ffmpeg/lib
 
 all: all-am
 
 .SUFFIXES:
 .SUFFIXES: .cpp .lo .o .obj

--- opencv-1.0.0/interfaces/swig/python/Makefile.am.orig	2008-07-30 17:45:39.874931035 -0700
+++ opencv-1.0.0/interfaces/swig/python/Makefile.am	2008-07-30 17:54:08.486284849 -0700
@@ -112,8 +112,8 @@
 _highgui_la_SOURCES    = _highgui.cpp pyhelpers.cpp pyhelpers.h \
                          highgui.i
 _highgui_la_CXXFLAGS     = $(PYTHON_CSPEC)
 _highgui_la_LIBADD       = $(top_builddir)/otherlibs/highgui/libhighgui.la
 _highgui_la_LDFLAGS      = -module -avoid-version -no-undefined $(PYTHON_LSPEC) \
-                           -L$(SWIG_PYTHON_LIBS)
+                           -L$(SWIG_PYTHON_LIBS) -Wl,-rpath,`rospack find ffmpeg`/ffmpeg/lib
 
 endif
