Index: CMakeLists.txt
===================================================================
--- CMakeLists.txt	(revision 1489)
+++ CMakeLists.txt	(working copy)
@@ -132,7 +132,7 @@
 set(BUILD_TESTS ON CACHE BOOL "Build tests")
 
 if(UNIX)
-    include(FindPkgConfig OPTIONAL)
+    include($ENV{ROS_ROOT}/core/rosbuild/FindPkgConfig.cmake OPTIONAL)
     include(CheckFunctionExists)
     include(CheckIncludeFile)
 endif()
@@ -170,7 +170,9 @@
             endforeach()
 
             foreach(P "${ALIAS_LIBRARY_DIRS}")
-                list(APPEND HIGHGUI_LIBRARY_DIRS ${${P}})
+	    	if (${${P}})
+                    list(APPEND HIGHGUI_LIBRARY_DIRS ${${P}})
+                endif(${${P}})
             endforeach()
 
             list(APPEND HIGHGUI_LIBRARIES ${${ALIAS_LIBRARIES}})
@@ -471,7 +473,8 @@
 set(CMAKE_INCLUDE_DIRS_CONFIGCMAKE "${CMAKE_INSTALL_PREFIX}/include/opencv")
 set(CMAKE_LIB_DIRS_CONFIGCMAKE "${CMAKE_INSTALL_PREFIX}/lib")
 
-exec_program(${CMAKE_COMMAND} ARGS "-E make_directory \"${CMAKE_BINARY_DIR}/unix-install/\"")
+#exec_program(${CMAKE_COMMAND} ARGS "-E make_directory \"${CMAKE_BINARY_DIR}/unix-install/\"")
+exec_program(mkdir ARGS "-p \"${CMAKE_BINARY_DIR}/unix-install/\"")
 configure_file("${CMAKE_SOURCE_DIR}/OpenCVConfig.cmake.in" "${CMAKE_BINARY_DIR}/unix-install/OpenCVConfig.cmake" IMMEDIATE)
 
 if(UNIX)
