diff --git FaceRecEm/common/src/b_FDSDK/SDK.h FaceRecEm/common/src/b_FDSDK/SDK.h
index 33742e5..9254971 100644
--- FaceRecEm/common/src/b_FDSDK/SDK.h
+++ FaceRecEm/common/src/b_FDSDK/SDK.h
@@ -29,6 +29,8 @@ extern "C" {
 /* ---- includes ----------------------------------------------------------- */
 
 #include "Types.h"
+#include <stddef.h>
+
 
 /* ---- related objects  --------------------------------------------------- */
 
@@ -41,7 +43,7 @@ struct btk_SDK;
 typedef struct btk_SDK* btk_HSDK;
 
 /** malloc function pointer */
-typedef void* ( *btk_fpMalloc )( u32 sizeA );
+typedef void* ( *btk_fpMalloc )( size_t sizeA );
 
 /** free function pointer */
 typedef void ( *btk_fpFree )( void* memPtrA );
