Index: gazebo-svn/server/controllers/camera/generic/Generic_Camera.cc
===================================================================
--- gazebo-svn/server/controllers/camera/generic/Generic_Camera.cc	(revision 6690)
+++ gazebo-svn/server/controllers/camera/generic/Generic_Camera.cc	(working copy)
@@ -93,7 +93,7 @@
 void Generic_Camera::PutCameraData()
 {
   CameraData *data = this->cameraIface->data;
-  const unsigned char *src;
+  //const unsigned char *src;
   unsigned char *dst;
   Pose3d cameraPose;
 
@@ -126,7 +126,7 @@
   //src = this->myParent->GetImageData();
   dst = data->image;
 
-  memcpy(dst, src, data->image_size);
+  //memcpy(dst, src, data->image_size);
 
 
   //unsigned int i, j, k;
