Index: plugins/rosrave/roslaser2d.cpp
===================================================================
--- plugins/rosrave/roslaser2d.cpp	(revision 609)
+++ plugins/rosrave/roslaser2d.cpp	(working copy)
@@ -123,7 +123,7 @@
 void ROSLaser2D::startsubscriptions()
 {
     // check if thread launched
-    ros::node* pnode = ros::node::instance();
+    ros::Node* pnode = ros::Node::instance();
 
     if( pnode && !pnode->checkMaster() ) {
         ros::fini();
@@ -137,7 +137,7 @@
         char strname[256] = "nohost";
         gethostname(strname, sizeof(strname));
         strcat(strname,"_openrave");
-        pnode = new ros::node(strname, ros::node::DONT_HANDLE_SIGINT|ros::node::ANONYMOUS_NAME);
+        pnode = new ros::Node(strname, ros::Node::DONT_HANDLE_SIGINT|ros::Node::ANONYMOUS_NAME);
     }
 
     if( pnode->checkMaster() ) {
@@ -149,7 +149,7 @@
 void ROSLaser2D::stopsubscriptions()
 {
     if( bSubscribed ) {
-        ros::node* pnode = ros::node::instance();
+        ros::Node* pnode = ros::Node::instance();
         if( pnode && pnode->checkMaster() ) {
             pnode->unsubscribe(scantopic.c_str());
             bSubscribed = false;
