Index: src/LinearMath/btQuaternion.h
===================================================================
--- src/LinearMath/btQuaternion.h	(revision 1458)
+++ src/LinearMath/btQuaternion.h	(working copy)
@@ -213,6 +213,12 @@
 		return s;
 	}
 
+  /**@brief Return the axis of the rotation represented by this quaternion */
+	btVector3 getAxis() const
+	{
+		btScalar s = btSqrt(btScalar(1.) - btPow(m_floats[3], 2.));
+		return btVector3(m_floats[0] / s, m_floats[1] / s, m_floats[2] / s);
+	}
 
   /**@brief Return the inverse of this quaternion */
 	btQuaternion inverse() const
