Index: src/LinearMath/btQuaternion.h
===================================================================
--- src/LinearMath/btQuaternion.h	(revision 1422)
+++ src/LinearMath/btQuaternion.h	(working copy)
@@ -198,7 +198,12 @@
 	{
 		btScalar s = btSqrt(length2() * q.length2());
 		assert(s != btScalar(0.0));
-		return btAcos(dot(q) / s);
+                btScalar retval;
+                if (dot(q) < 0) // Take care of long angle case see http://en.wikipedia.org/wiki/Slerp
+                  retval = btAcos(GEN_clamped(dot(-q) / s, (btScalar)-1.0, (btScalar)1.0)); //Clamped to prevent floating point precision crashes
+                else 
+                  retval = btAcos(GEN_clamped(dot(q) / s, (btScalar)-1.0, (btScalar)1.0)); //Clamped to prevent floating point precision crashes
+		return retval;
 	}
   /**@brief Return the angle of rotation represented by this quaternion */
 	btScalar getAngle() const 
