*** bullet-2.71/src/LinearMath/btTransform.h.original	2008-09-16 14:53:44.985742529 -0700
--- bullet-2.71/src/LinearMath/btTransform.h	2008-09-16 14:54:48.335742901 -0700
*************** public:
*** 82,87 ****
--- 82,91 ----
  		return (*this)(x);
  	}
  
+ 	SIMD_FORCE_INLINE btQuaternion operator*(const btQuaternion& q) const
+ 	{
+ 		return getRotation() * q;
+ 	}
  	SIMD_FORCE_INLINE btMatrix3x3&       getBasis()          { return m_basis; }
  	SIMD_FORCE_INLINE const btMatrix3x3& getBasis()    const { return m_basis; }
  
