diff -Naur kdl_export/src/frames.cpp kdl_new/src/frames.cpp
--- src/frames.cpp	2008-08-19 04:50:49.000000000 -0700
+++ src/frames.cpp	2008-11-11 10:07:42.995743193 -0800
@@ -194,6 +194,14 @@
 
 }
 
+Rotation Rotation::Quaternion(double x,double y,double z, double w)
+    {
+        double x2, y2, z2, w2;
+	x2 = x*x;  y2 = y*y; z2 = z*z;  w2 = w*w;
+        return Rotation(w2+x2-y2-z2, 2*x*y-2*w*z, 2*x*z+2*w*y,
+			2*x*y+2*w*z, w2-x2+y2-z2, 2*y*z-2*w*x,
+			2*x*z-2*w*y, 2*y*z+2*w*x, w2-x2-y2+z2);
+    }
 
 Rotation Rotation::RPY(double roll,double pitch,double yaw)
     {
diff -Naur kdl_export/src/frames.hpp kdl_new/src/frames.hpp
--- src/frames.hpp	2008-08-19 04:50:49.000000000 -0700
+++ src/frames.hpp	2008-11-11 08:33:31.005807894 -0800
@@ -390,6 +390,8 @@
     //!  (-PI <= alfa <= PI)
     void GetEulerZYZ(double& alfa,double& beta,double& gamma) const;
 
+    //! Sets the value of this object to a rotation specified with Quaternion convention
+    static Rotation Quaternion(double x,double y,double z, double w);
 
     //! Sets the value of this object to a rotation specified with RPY convention:
     //! first rotate around X with roll, then around the
