Index: gridfastslam/gridslamprocessor.cpp
===================================================================
--- gridfastslam/gridslamprocessor.cpp	(revision 39)
+++ gridfastslam/gridslamprocessor.cpp	(working copy)
@@ -402,6 +402,13 @@
 	plainReading[i]=reading[i];
       }
       m_infoStream << "m_count " << m_count << endl;
+
+      RangeReading* reading_copy = 
+              new RangeReading(reading.size(),
+                               &(reading[0]),
+                               static_cast<const RangeSensor*>(reading.getSensor()),
+                               reading.getTime());
+
       if (m_count>0){
 	scanMatch(plainReading);
 	if (m_outputStream.is_open()){
@@ -432,7 +439,7 @@
 	  m_outputStream << setiosflags(ios::fixed) << setprecision(6);
 	  m_outputStream << "NEFF " << m_neff << endl;
 	}
-	resample(plainReading, adaptParticles);
+	resample(plainReading, adaptParticles, reading_copy);
 	
       } else {
 	m_infoStream << "Registering First Scan"<< endl;
@@ -443,9 +450,9 @@
 	  
 	  // cyr: not needed anymore, particles refer to the root in the beginning!
 	  TNode* node=new	TNode(it->pose, 0., it->node,  0);
-	  node->reading=0;
+	  //node->reading=0;
+          node->reading = reading_copy;
 	  it->node=node;
-	  
 	}
       }
       //		cerr  << "Tree: normalizing, resetting and propagating weights at the end..." ;
Index: gridfastslam/gridslamprocessor.hxx
===================================================================
--- gridfastslam/gridslamprocessor.hxx	(revision 39)
+++ gridfastslam/gridslamprocessor.hxx	(working copy)
@@ -67,7 +67,7 @@
   
 }
 
-inline bool GridSlamProcessor::resample(const double* plainReading, int adaptSize, const RangeReading* ){
+inline bool GridSlamProcessor::resample(const double* plainReading, int adaptSize, const RangeReading* reading){
   
   bool hasResampled = false;
   
@@ -112,7 +112,7 @@
       TNode* oldNode=oldGeneration[m_indexes[i]];
       //			cerr << i << "->" << m_indexes[i] << "B("<<oldNode->childs <<") ";
       node=new	TNode(p.pose, 0, oldNode, 0);
-      node->reading=0;
+      node->reading=reading;
       //			cerr << "A("<<node->parent->childs <<") " <<endl;
       
       temp.push_back(p);
@@ -155,7 +155,8 @@
       TNode* node=0;
       node=new TNode(it->pose, 0.0, *node_it, 0);
       
-      node->reading=0;
+      //node->reading=0;
+      node->reading=reading;
       it->node=node;
 
       //END: BUILDING TREE
Index: scanmatcher/scanmatcher.h
===================================================================
--- scanmatcher/scanmatcher.h	(revision 39)
+++ scanmatcher/scanmatcher.h	(working copy)
@@ -7,7 +7,7 @@
 #include <utils/stat.h>
 #include <iostream>
 #include <utils/gvalues.h>
-#define LASER_MAXBEAMS 1024
+#define LASER_MAXBEAMS 2048
 
 namespace GMapping {
 
Index: particlefilter/particlefilter.h
===================================================================
--- particlefilter/particlefilter.h	(revision 39)
+++ particlefilter/particlefilter.h	(working copy)
@@ -1,6 +1,7 @@
 #ifndef PARTICLEFILTER_H
 #define PARTICLEFILTER_H
 #include <stdlib.h>
+#include <float.h>
 #include <sys/types.h>
 #include<vector>
 #include<utility>
@@ -23,7 +24,8 @@
 template <class OutputIterator, class Iterator>
 double toNormalForm(OutputIterator& out, const Iterator & begin, const Iterator & end){
 	//determine the maximum
-	double lmax=-MAXDOUBLE;
+	//double lmax=-MAXDOUBLE;
+	double lmax=-DBL_MAX;
 	for (Iterator it=begin; it!=end; it++){
 		lmax=lmax>((double)(*it))? lmax: (double)(*it);
 	}
--- configfile/configfile.cpp.old	2008-11-13 01:38:20.000000000 -0800
+++ configfile/configfile.cpp	2008-11-13 01:38:41.000000000 -0800
@@ -30,6 +30,7 @@
 #include <sstream>
 #include <iostream>
 #include <ctype.h>
+#include <cstdlib>
 
 namespace GMapping{
 using namespace std;
--- configfile/configfile_test.cpp.old	2008-11-13 01:37:49.000000000 -0800
+++ configfile/configfile_test.cpp	2008-11-13 01:39:05.000000000 -0800
@@ -23,6 +23,7 @@
 
 
 #include <iostream>
+#include <cstdlib>
 #include "configfile.h"
 
 using namespace std;
--- ./gridfastslam/gfs2log.cpp.old	2008-11-13 01:47:21.000000000 -0800
+++ ./gridfastslam/gfs2log.cpp	2008-11-13 01:48:14.000000000 -0800
@@ -3,6 +3,7 @@
 #include <sstream>
 #include <vector>
 #include <list>
+#include <cstring>
 #include <utils/point.h>
 #include "gfsreader.h"
 
@@ -12,7 +13,7 @@
 using namespace GMapping;
 using namespace GMapping::GFSReader;
 
-int main (unsigned int argc, const char * const * argv){
+int main (int argc, const char * const * argv){
 	if (argc<3){
 		cout << "usage gfs2log [-err] [-neff] [-part] [-odom] <infilename> <outfilename>" << endl;
 		cout << "  -odom : dump raw odometry in ODOM message instead of inpolated corrected one" << endl;
--- ./gridfastslam/gfs2neff.cpp.old	2008-11-13 01:49:27.000000000 -0800
+++ ./gridfastslam/gfs2neff.cpp	2008-11-13 01:49:44.000000000 -0800
@@ -4,7 +4,7 @@
 
 using namespace std;
 
-int main(unsigned int argc, const char* const *argv){
+int main(int argc, const char* const *argv){
 	if (argc<3){
 		cout << "usage gfs2neff <infilename> <nefffilename>" << endl;
 		return -1;
--- ./gridfastslam/gfs2rec.cpp.old	2008-11-13 01:48:35.000000000 -0800
+++ ./gridfastslam/gfs2rec.cpp	2008-11-13 01:49:05.000000000 -0800
@@ -3,6 +3,7 @@
 #include <sstream>
 #include <vector>
 #include <list>
+#include <cstring>
 #include <utils/point.h>
 
 #define MAX_LINE_LENGHT (1000000)
@@ -368,7 +369,7 @@
 
 
 
-int main (unsigned int argc, const char * const * argv){
+int main ( int argc, const char * const * argv){
 	if (argc<3){
 		cout << "usage gfs2rec [-err] <infilename> <outfilename>" << endl;
 		return -1;
--- ./gridfastslam/gfsreader.cpp.old	2008-11-13 01:46:40.000000000 -0800
+++ ./gridfastslam/gfsreader.cpp	2008-11-13 01:46:57.000000000 -0800
@@ -1,6 +1,7 @@
 #include "gfsreader.h"
 #include <iomanip>
 #include <limits>
+#include <cstring>
 
 namespace  GMapping { 
 
--- ./gui/gfs2img.cpp.old	2008-11-13 01:50:34.000000000 -0800
+++ ./gui/gfs2img.cpp	2008-11-13 01:50:56.000000000 -0800
@@ -1,4 +1,5 @@
 #include <limits.h>
+#include <cstdlib>
 #include <scanmatcher/scanmatcher.h>
 #include <gridfastslam/gfsreader.h>
 #include <qpixmap.h>
--- ./log/carmenconfiguration.cpp.old	2008-11-13 01:34:00.000000000 -0800
+++ ./log/carmenconfiguration.cpp	2008-11-13 01:34:23.000000000 -0800
@@ -2,6 +2,7 @@
 #include <iostream>
 #include <sstream>
 #include <assert.h>
+#include <cstdlib>
 #include <sys/types.h>
 #include <sensor_odometry/odometrysensor.h>
 #include <sensor_range/rangesensor.h>
--- ./log/scanstudio2carmen.cpp.old	2008-11-13 01:36:00.000000000 -0800
+++ ./log/scanstudio2carmen.cpp	2008-11-13 01:36:11.000000000 -0800
@@ -1,6 +1,7 @@
 #include <iostream>
 #include <fstream>
 #include <sstream>
+#include <cstdlib>
 #include <assert.h>
 #include <utils/point.h>
 
--- ./log/rdk2carmen.cpp.old	2008-11-13 01:36:25.000000000 -0800
+++ ./log/rdk2carmen.cpp	2008-11-13 01:36:51.000000000 -0800
@@ -1,5 +1,6 @@
 #include <fstream>
 #include <iostream>
+#include <cstdlib>
 #include <log/carmenconfiguration.h>
 #include <log/sensorlog.h>
 
@@ -7,7 +8,7 @@
 using namespace std;
 using namespace GMapping;
 
-int main(char argc, char ** argv){
+int main(int argc, char ** argv){
 	if (argc<2){
 		cerr << "usage "<<argv[0]<<" <filename> <outfilename>" << endl;
 		cerr << "or "<<argv[0]<<" <filename> for standard output" << endl;
--- ./log/log_test.cpp.old	2008-11-13 01:35:19.000000000 -0800
+++ ./log/log_test.cpp	2008-11-13 01:35:34.000000000 -0800
@@ -1,5 +1,6 @@
 #include <fstream>
 #include <iostream>
+#include <cstdlib>
 #include <log/carmenconfiguration.h>
 #include <log/sensorlog.h>
 
@@ -7,7 +8,7 @@
 using namespace std;
 using namespace GMapping;
 
-int main(char argc, char ** argv){
+int main(int argc, char ** argv){
 	if (argc<2){
 		cout << "usage log_test <filename>" << endl;
 		exit (-1);
--- ./log/log_plot.cpp.old	2008-11-13 01:34:40.000000000 -0800
+++ ./log/log_plot.cpp	2008-11-13 01:35:07.000000000 -0800
@@ -1,5 +1,6 @@
 #include <fstream>
 #include <iostream>
+#include <cstdlib>
 #include <sys/types.h>
 #include <log/carmenconfiguration.h>
 #include <log/sensorlog.h>
@@ -8,7 +9,7 @@
 using namespace std;
 using namespace GMapping;
 
-int main(char argc, char ** argv){
+int main(int argc, char ** argv){
   double maxrange=2.;
 	if (argc<2){
 		cout << "usage log_plot <filename> | gnuplot" << endl;
--- ./scanmatcher/scanmatcher.cpp.old	2008-11-13 01:43:08.000000000 -0800
+++ ./scanmatcher/scanmatcher.cpp	2008-11-13 01:44:25.000000000 -0800
@@ -1,5 +1,7 @@
 #include <list>
 #include <iostream>
+#include <cstring>
+#include <limits>
 
 #include "scanmatcher.h"
 #include "gridlinetraversal.h"
--- ./scanmatcher/gridlinetraversal.h.old	2008-11-13 01:43:37.000000000 -0800
+++ ./scanmatcher/gridlinetraversal.h	2008-11-13 01:44:06.000000000 -0800
@@ -1,6 +1,7 @@
 #ifndef GRIDLINETRAVERSAL_H
 #define GRIDLINETRAVERSAL_H
 
+#include <cstdlib>
 #include <utils/point.h>
 
 namespace GMapping {
--- ./scanmatcher/icptest.cpp.old	2008-11-13 01:45:40.000000000 -0800
+++ ./scanmatcher/icptest.cpp	2008-11-13 01:45:59.000000000 -0800
@@ -1,5 +1,6 @@
 #include <iostream>
 #include <fstream>
+#include <cstdlib>
 
 #include <list>
 #include "icp.h"
--- ./scanmatcher/scanmatch_test.cpp.old	2008-11-13 01:44:57.000000000 -0800
+++ ./scanmatcher/scanmatch_test.cpp	2008-11-13 01:45:14.000000000 -0800
@@ -1,4 +1,5 @@
 
+#include <cstdlib>
 #include <fstream>
 #include <iostream>
 #include <log/carmenconfiguration.h>
--- ./sensor/sensor_range/rangereading.cpp.old	2008-11-13 01:33:11.000000000 -0800
+++ ./sensor/sensor_range/rangereading.cpp	2008-11-13 01:33:30.000000000 -0800
@@ -1,3 +1,4 @@
+#include <limits>
 #include <iostream>
 #include <assert.h>
 #include <sys/types.h>
