Index: src/filter/kalmanfilter.h
===================================================================
--- src/filter/kalmanfilter.h	(revision 29726)
+++ src/filter/kalmanfilter.h	(working copy)
@@ -79,9 +79,9 @@
   struct MeasUpdateVariables
   {
     Matrix _S;
-    Matrix _postHT;
     Matrix _K;
     ColumnVector _innov;
+    Matrix _postHT;
     MeasUpdateVariables() {};
     MeasUpdateVariables(unsigned int meas_dimension, unsigned int state_dimension):
       _S(meas_dimension,meas_dimension) 
Index: src/filter/extendedkalmanfilter.cpp
===================================================================
--- src/filter/extendedkalmanfilter.cpp	(revision 29726)
+++ src/filter/extendedkalmanfilter.cpp	(working copy)
@@ -90,6 +90,7 @@
 {
   // allocate measurement for z.rows() if needed
   AllocateMeasModelExt(z.rows());
+  _x = _post->ExpectedValueGet();
 
   (_mapMeasUpdateVariablesExt_it->second)._Z = ((AnalyticMeas*)measmodel)->PredictionGet(s,_x);
   (_mapMeasUpdateVariablesExt_it->second)._H = ((AnalyticMeas*)measmodel)->df_dxGet(s,_x);
