Index: ode/src/quickstep.cpp
===================================================================
--- ode/src/quickstep.cpp	(revision 1550)
+++ ode/src/quickstep.cpp	(working copy)
@@ -66,6 +66,9 @@
 
 #define RANDOMLY_REORDER_CONSTRAINTS 1
 
+
+//#define TIMING
+
 //****************************************************************************
 // special matrix multipliers
 
Index: ode/src/ode.cpp
===================================================================
--- ode/src/ode.cpp	(revision 1550)
+++ ode/src/ode.cpp	(working copy)
@@ -1520,7 +1520,7 @@
   w->adis.angular_average_threshold = REAL(0.01)*REAL(0.01);	// (magnitude squared)
   w->adis.linear_average_threshold = REAL(0.01)*REAL(0.01);		// (magnitude squared)
 
-  w->qs.num_iterations = 20;
+  w->qs.num_iterations = 5;
   w->qs.w = REAL(1.3);
 
   w->contactp.max_vel = dInfinity;
