--- /home/mquigley/Desktop/KNI_3.9.2/include/KNI_LM/lmBase.h	2007-04-12 05:52:06.000000000 -0700
+++ include/KNI_LM/lmBase.h	2008-06-26 21:16:09.000000000 -0700
@@ -242,6 +242,10 @@
     /// This is in the case you want to initialize it manually
     /// @note If you do not call it, moveRobotLinearTo() will do it for you automatically
     void initLM();
+    void getSplineCoeff(int steps, double *timearray, double *encoderarray,
+                        double *arr_p1, double *arr_p2, double *arr_p3, double *arr_p4);
+    bool checkJointSpeeds(std::vector<int> lastsolution, 
+                          std::vector<int> solution, double time);
 
 	/// @param wait has to be true with new implementation of movLM2P
     void movLM(double X, double Y, double Z, 
