Index: src/tree.hpp
===================================================================
--- src/tree.hpp	(revision 30453)
+++ src/tree.hpp	(working copy)
@@ -142,7 +142,9 @@
          */
         SegmentMap::const_iterator getSegment(const std::string& segment_name)const
         {
-            return segments.find(segment_name);
+          SegmentMap::const_iterator it = segments.find(segment_name);
+          if (it == segments.end()) throw seg_name_ex;
+          else return it;
         };
         /**
          * Request the root segment of the tree
@@ -173,6 +175,13 @@
 
         virtual ~Tree(){};
 
+
+      class segment_name_exception: public std::exception{
+        virtual const char* what() const throw(){
+          return "Segment Name excption";}
+      } seg_name_ex;
+
+
     };
 }
 #endif
