Index: src/al/ethercat_AL.cxx
===================================================================
--- src/al/ethercat_AL.cxx	(revision 36)
+++ src/al/ethercat_AL.cxx	(working copy)
@@ -177,7 +177,7 @@
       else { // No such slave found...
 	log(EC_LOG_WARNING, "EC_AL Warning: No such slave in db, creating dummy slave\n");
 	// Create slave handler
-	m_slave_handler[i] = new EtherCAT_SlaveHandler(adp2ringpos(adp),productcode,revision,adp2ringpos(adp),NULL,NULL);
+	m_slave_handler[i] = new EtherCAT_SlaveHandler(adp2ringpos(adp),productcode,revision,(i+1),NULL,NULL);
 	log(EC_LOG_INFO, "AL creating SlaveHandler: pos=%d, Product Code=0x%x, rev=0x%x\n", 
 	       adp2ringpos(adp),productcode,revision);
       }

--- include/al/ethercat_slave_conf.h	2008-07-20 15:13:52.516068789 -0700
+++ include/al/ethercat_slave_conf.h	2008-07-21 14:56:25.114819033 -0700
@@ -160,8 +160,10 @@
   EC_FixedStationAddress get_station_address() const {return m_station_address; };
   /// Get FMMU config to be written when going to preop
   const EtherCAT_FMMU_Config * get_fmmu_config() const {return m_fmmu_config; };
+  void set_fmmu_config(EtherCAT_FMMU_Config *new_config) { m_fmmu_config = new_config; };
   /// Get PD Configuration (SMS)
   const EtherCAT_PD_Config * get_pd_config() const {return m_pd_config; };
+  void set_pd_config(EtherCAT_PD_Config *new_config) { m_pd_config = new_config; };
   /// Get configuration of Sync Man channel 0 and 1 for MBX
   /** @return pointer to config in case of complex slave, or NULL for
       a simple slave
