Index: tests/posix/eval_test.cpp
===================================================================
--- tests/posix/eval_test.cpp	(revision 36)
+++ tests/posix/eval_test.cpp	(working copy)
@@ -40,6 +40,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <dll/ethercat_dll.h>
+#include <unistd.h>
 #include <dll/ethercat_frame.h>
 #include <al/ethercat_master.h>
 #include <al/ethercat_slave_handler.h>
Index: tests/rtnet/eval_test.cpp
===================================================================
--- tests/rtnet/eval_test.cpp	(revision 36)
+++ tests/rtnet/eval_test.cpp	(working copy)
@@ -40,6 +40,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <dll/ethercat_dll.h>
+#include <unistd.h>
 #include <dll/ethercat_frame.h>
 #include <al/ethercat_master.h>
 #include <al/ethercat_slave_handler.h>
