--- /home/mquigley/Desktop/KNI_3.9.2/src/Base/cdlCOM.cpp	2007-04-12 05:52:06.000000000 -0700
+++ src/Base/cdlCOM.cpp	2008-06-06 09:56:43.000000000 -0700
@@ -169,11 +169,18 @@
 
 	errno = 0;
 
-	strncpy(name, "/dev/ttyS ", 11);
-	name[9] = digit(ccd.port);
-	prtHdl = ::open(name, O_RDWR | O_NOCTTY | O_NDELAY| O_NONBLOCK);
+  if (!ccd.port_str.length())
+  {
+  	strncpy(name, "/dev/ttyS ", 11);
+  	name[9] = digit(ccd.port);
+    deviceName = name;
+  }
+  else
+    deviceName = ccd.port_str; 
 
-	_deviceName = name;
+	prtHdl = ::open(deviceName.c_str(), O_RDWR | O_NOCTTY | O_NDELAY| O_NONBLOCK);
+
+	_deviceName = deviceName;
 
 	if (prtHdl < 0) {
 		throw CannotOpenPortException(_deviceName, strerror(errno));
@@ -289,7 +296,6 @@
 	}
 
 	_prtHdl = prtHdl;
-	_deviceName = deviceName;
 	_ccd = ccd;
 	_oto = oto;
 }
