--- src/LM/Makefile	2007-04-25 23:18:24.000000000 -0700
+++ /home/mquigley/stair/pr3/3rdparty/drivers/kni-3.9.2/KNI_3.9.2/src/LM/Makefile	2008-06-04 18:15:01.000000000 -0700
@@ -17,7 +17,7 @@
 #===========================================================================
 
 CXX		=	g++
-CXXFLAGS	+=	-fPIC -g -O2 -Wall
+CXXFLAGS	+=	-fPIC -g -O2 -Wall `rosboost-cfg --cflags`
 
 MAJOR_VERSION	=	3.9
 MINOR_VERSION	=	0
