--- cvcap_ffmpeg_orig.cpp	2008-03-18 00:08:52.000000000 -0700
+++ cvcap_ffmpeg.cpp	2008-03-18 00:06:46.000000000 -0700
@@ -40,6 +40,8 @@
 //M*/
 
 #include "_highgui.h"
+#define INT64_C
+#define __STDC_CONSTANT_MACROS
 
 
 extern "C" {
@@ -781,7 +783,7 @@
 
 	if (!(mywriter->fmt->flags & AVFMT_NOFILE)) {
 		/* close the output file */
-		url_fclose(&mywriter->oc->pb);
+		url_fclose(mywriter->oc->pb);
 	}
 
 	/* free the stream */
