Index: gazebo-svn/libgazebo/Iface.cc
===================================================================
--- gazebo-svn/libgazebo/Iface.cc	(revision 6690)
+++ gazebo-svn/libgazebo/Iface.cc	(working copy)
@@ -55,6 +55,8 @@
 GZ_REGISTER_IFACE("factory", FactoryIface);
 GZ_REGISTER_IFACE("gripper", GripperIface);
 GZ_REGISTER_IFACE("actarray", ActarrayIface);
+GZ_REGISTER_IFACE("pr2array", PR2ArrayIface);
+GZ_REGISTER_IFACE("pr2gripper", PR2GripperIface);
 GZ_REGISTER_IFACE("ptz", PTZIface);
 GZ_REGISTER_IFACE("stereocamera", StereoCameraIface);
 
