Index: gazebo-svn/server/physics/SliderJoint.cc
===================================================================
--- gazebo-svn/server/physics/SliderJoint.cc	(revision 6690)
+++ gazebo-svn/server/physics/SliderJoint.cc	(working copy)
@@ -35,6 +35,8 @@
     : Joint()
 {
   this->jointId = dJointCreateSlider( worldId, NULL );
+  this->type = Joint::SLIDER;
+  fprintf(stderr," slider jointId %d\n",this->jointId);
 }
 
 
